usr/share/rear/lib/layout-functions.sh

```diff
@@ -693,7 +693,18 @@ is_disk_a_pv() {
 }
 
 function is_multipath_path {
-    [ "$1" ] && type multipath &>/dev/null && multipath -c /dev/$1 &>/dev/null
+    # Return 'false' if there is no device as argument:
+    test "$1" || return 1
+    # Return 'false' if there is no multipath command:
+    type multipath &>/dev/null || return 1
+    # Return 'false' if multipath is not used, see https://github.com/rear/rear/issues/2298
+    # Because "multipath -l" always returns zero exit code we check if it has real output via grep -q '[[:alnum:]]'
+    # so that no "multipath -l" output could clutter the log (the "multipath -l" output is irrelevant here)
+    # in contrast to e.g. test "$( multipath -l )" that would falsely succeed with blank output
+    # and the output would appear in the log in 'set -x' debugscript mode:
+    multipath -l | grep -q '[[:alnum:]]' || return 1
```

Comment:
This reintroduces a performance problem that was fixed in PR #2034. `multipath -l` scans all devices and the time it takes is proportional to their number, so the total time spent on all this is quadratic in the number of devices. For 1600 devices it used to take 300 s, now it takes 4445 s. See https://github.com/rear/rear/pull/2597#issuecomment-920266696, https://github.com/rear/rear/pull/2597#issuecomment-814936492.

Reply:
First and foremost:
I am not at all a multipath expert
but nevertheless somehow it is mostly me
who has to deal with bad multipath code
and I get step by step more and more annoyed about that.

I neither see how I _re_-introduced the performance problem
nor do I see how the `multipath -l` performance problem
was fixed by https://github.com/rear/rear/pull/2034
because as far as I see its changes do not contain `multipath -l`
but as multipath noob I can easily miss imortant things here.

Furthermore we use `multipath -l` at other places in our code
and as far as I know this is the only simple and reliably working
commad to find out in general whether or not multipath is used at all.
So I introduced the `multipath -l` test to avoid that the subsequent
wrong working `multipath -c` when multipath is not used
can mess up all on "normal" (i.e. non-multipath) systems.

This pull request was even approved by @schabrolles
who was our multipath expert who maintained it so well in ReaR.
But unfortunately he gets no longer time from his employer IBM
to further maintain ReaR which proves things are OK as is for IBM.
I mean:
It is those big systems with multipath where IBM makes money with
but when IBM does no longer care about ReaR
then IBM gets what they pay for.

Reply:
Hi @jsmeix , sorry for the wrong formulation. The problem is new, but the problem fixed #2034 was also due to quadratic time complexity in multipath code, so the result is indistinguishable (inordinate amount of time spent creating the rescue image when there are lots of multipath devices) and for the end user (or a quality engineer) it looks like a regression.

Concerning
> I am not at all a multipath expert
> but nevertheless somehow it is mostly me
> who has to deal with bad multipath code
> and I get step by step more and more annoyed about that.

I am also not a multipath expert, but I may slowly become one, because I also have to deal with multipath issues in ReaR from time to time. So, feel free to assign multipath issues to me, or tag me for review of changes.

Concerning this particular change: do you have a test case for the problem that it fixed (#2298)? I will need to verify that any fix I am going to make does not reintroduce the problem.

Reply:
Well, I'm still in vacation ;) but I had a quick look on it. If I understand well, the main main issue here is the fact that `multipath -l` is listing all the devices and can some time when a huge number of devices are connected to the server. And because this function will be executed for **each** device it's even worse.

My 2 cents:

1) @pcahyna, could you try on your system with a lot of devices if the following command is quicker than `multipath -l` 
`dmsetup ls --target multipath`

2) The other other option would be to remove this test from "the loop" for each device.
May be create a first test that create a variable like `multipath_device_count=$(dmsetup ls --target multipath | wc -l)`.
This variable could be use to enter into specific code to treat multipath devices.

Reply:
@pcahyna 
thank you for your explanatory reply.

I have no (automated) test cases.
I only test manually (I know, I know ... ;-)

But I have an idea how to generically solve
the underlying generic issue:

The underlying generic issue is that
no multipath code should be run when multipath is not used.
I.e. on non-multipath systems all multipath code should be skipped.

To do that it is sufficient to run a test only once
to find out whether or not multipath is used
and remember the result (in a global variable)
and then only use that gloval variable.

So we could have a new function like
(just off the top of my head to show the idea - not at all tested):
```
function is_multipath_used {
    # 'multipath -l' is the only simple and reliably working commad
    # to find out in general whether or not multipath is used at all.
    # But 'multipath -l' scans all devices and the time it takes is proportional
    # to their number so that time becomes rather long (seconds up to minutes)
    # if 'multipath -l' was called for each one of hundreds or thousands of devices.
    # So we call 'multipath -l' only once and remember the result
    # in a global variable and then only use that global variable
    # so we can call is_multipath_used very many times as often needed.
    is_true $MULTIPATH_IS_USED && return 0
    is_false $MULTIPATH_IS_USED && return 1
    # When MULTIPATH_IS_USED has neither a true nor false value set it and return accordingly.
    # Because "multipath -l" always returns zero exit code we check if it has real output via grep -q '[[:alnum:]]'
    # so that no "multipath -l" output could clutter the log (the "multipath -l" output is irrelevant here)
    # in contrast to e.g. test "$( multipath -l )" that would falsely succeed with blank output
    # and the output would appear in the log in 'set -x' debugscript mode:
    if multipath -l | grep -q '[[:alnum:]]' ; then
        MULTIPATH_IS_USED='yes'
        return 0
    else
        MULTIPATH_IS_USED='no'
        return 1
    fi
}
```
plus
```
MULTIPATH_IS_USED=''
```
in default.conf together with an explanatory comment
to provide "final power to the user".

Reply:
Hi @schabrolles , my ideas for making the test quicker and more reliable are described in https://github.com/rear/rear/pull/2597#issuecomment-815038677 (test whether `lsblk --nodeps -o fstype /dev/sd...` shows `mpath_member` ) and https://github.com/rear/rear/pull/2597#issuecomment-816587171 (test whether there is a `/sys/block/*/holders/*` symlink pointing to the multipath device). I suppose `dmsetup` would work as well.

Reply:
@schabrolles 
happy to hear from you again!

No need at all to say sorry that you can no longer maintain ReaR.
It is IBM who should say sorry that they do not support ReaR.
SUSE supports ReaR.
Red Hat supports ReaR.
But IBM itself does not (directly) support it.
Same for HP.
Same for all other "big server" manufacturers.

I didn't see your posting because I was about writing mine.
Your proposal "2." is essentially what also I suggest.

Reply:
@schabrolles 
of course you and your family have way much higher priority
than any piece of software could ever have
(living humans versus "dead" programs).
Of course I don't expect any further work for ReaR from you.
I do much appreciate all what you contributed.
So just enjoy your vacation - i.e. relax and recover!
I wish you and your family all the very best.

Reply:
@pcahyna 
if you think something like a is_multipath_used function is useful
it could of couse implement any test that is appropriate,
perhaps even several tests could be needed
to make it work also on old systems where e.g.
`lsblk --nodeps -o fstype /dev/sd...` may not yet work
so e.g. `multipath -l` might be used as final fallback
(or not at all when other test work better in general).

Reply:
@pcahyna
please have a look at https://github.com/rear/rear/pull/2708

In particular if you know a better method than `multipath -l`
which we can use in the new function is_multipath_used.